core/src/main/java/org/mskcc/cbio/portal/scripts/ImportResourceData.java

```diff
@@ -0,0 +1,555 @@
+    }
+
+    private int findSampleTypeColumn(Map<String, Integer> headerIndexMap) {
+        return findColumnIndexInHeaders(SAMPLE_TYPE_COLUMN_NAME, headerIndexMap);
+    }
+
+    private int findColumnIndexInHeaders(String columnHeader, Map<String, Integer> headerIndexMap) {
+        return headerIndexMap.getOrDefault(columnHeader, -1);
+    }
+
+    private int addPatientToDatabase(String patientId) throws Exception {
+        int internalPatientId = -1;
+        if (validPatientId(patientId)) {
+            Patient patient = DaoPatient.getPatientByCancerStudyAndPatientId(cancerStudy.getInternalId(), patientId);
+            // other validations:
+            // in case of PATIENT data import, there are some special checks:
+            if (getResourceType() == ResourceType.PATIENT) {
+                // if resource data is already there, then something has gone wrong (e.g.
+                // patient is duplicated in file), abort:
+                if (patient != null
+                        && DaoResourceData.getDataByPatientId(cancerStudy.getInternalId(), patientId).size() > 0) {
+                    throw new RuntimeException(
+                            "Something has gone wrong. Patient " + patientId + " already has resource data loaded.");
+                }
+                // if patient is duplicated, abort as well in this case:
+                if (!patientIds.add(patientId)) {
+                    throw new RuntimeException("Error. Patient " + patientId + " found to be duplicated in your file.");
+                }
+            }
+
+            if (patient != null) {
+                // in all cases (SAMPLE, PATIENT, or STUDY) this can be expected, so
+                // just fetch it:
+                internalPatientId = patient.getInternalId();
+            } else {
+                // in case of PATIENT data import and patient == null :
+                if (getResourceType() == ResourceType.PATIENT) {
+                    // not finding the patient it unexpected (as SAMPLE data import should always
+                    // precede it), but
+                    // can happen when this patient does not have any samples for example. In any
+                    // case, warn about it:
+                    ProgressMonitor.logWarning("Patient " + patientId
+                            + " being added for the first time. Apparently this patient was not in the samples file, or the samples file is not yet loaded (should be loaded before this one)");
+                }
+
+                patient = new Patient(cancerStudy, patientId);
+                internalPatientId = DaoPatient.addPatient(patient);
+            }
+        }
+        return internalPatientId;
+    }
+
+    private int addSampleToDatabase(String sampleId, String[] fields, Map<String, Integer> headerIndexMap)
+            throws Exception {
+        int sampleTypeIndex = findSampleTypeColumn(headerIndexMap);
```

Comment:
Are we going to have `sampleType`(`SAMPLE_TYPE` column) in data file?

Reply:
Yes, IMO, if someone introduces a new sample in the resource data file, `SAMPLE_TYPE` can be an optional column in the data file.